vortex-array/src/array/sparse/canonical.rs

```diff
@@ -29,75 +31,72 @@ impl IntoCanonical for SparseArray {
                     values.validity(),
                     &indices,
                     self.len(),
-                    self.fill_value()
+                    self.fill_value(),
+                    self.dtype().nullability()
                 )
             })
         }
     }
 }
 
 fn canonicalize_sparse_bools(
-    values_array: BoolArray,
+    values: BoolArray,
     indices: &[usize],
     len: usize,
     fill_value: &ScalarValue,
+    nullability: Nullability,
 ) -> VortexResult<Canonical> {
-    let fill_bool = if fill_value.is_null() {
-        false
+    let (fill_bool, validity) = if fill_value.is_null() {
+        (false, Validity::AllInvalid)
     } else {
-        fill_value.try_into()?
+        (
+            fill_value.try_into()?,
+            if nullability == Nullability::NonNullable {
+                Validity::NonNullable
+            } else {
+                Validity::AllValid
+            },
+        )
     };
 
-    let values_validity = values_array.validity();
-    let values = values_array.boolean_buffer();
-
-    // pre-fill both values and validity based on fill_value
-    // this optimizes performance for the common case where indices.len() < len / 2
-    let mut flat_bools = BooleanBufferBuilder::new(len);
-    flat_bools.append_n(len, fill_bool);
-    let mut validity_buffer = BooleanBufferBuilder::new(len);
-    validity_buffer.append_n(len, !fill_value.is_null());
-
-    // patch in the actual values and validity
-    for (i, idx) in indices.iter().enumerate() {
-        flat_bools.set_bit(*idx, values.value(i));
-        validity_buffer.set_bit(*idx, values_validity.is_valid(i));
-    }
-
-    BoolArray::try_new(
-        flat_bools.finish(),
-        Validity::from(validity_buffer.finish()),
-    )
-    .map(Canonical::Bool)
+    let bools = BoolArray::try_new(
+        if fill_bool {
+            BooleanBuffer::new_set(len)
+        } else {
+            BooleanBuffer::new_unset(len)
+        },
+        validity,
+    )?;
+    let patched = bools.patch(indices, values)?;
+    Ok(Canonical::Bool(patched))
 }
 
 fn canonicalize_sparse_primitives<
-    T: NativePType + for<'a> TryFrom<&'a ScalarValue, Error = VortexError>,
+    T: NativePType + for<'a> TryFrom<&'a ScalarValue, Error = VortexError> + ArrowNativeType,
```

Comment:
why does `PrimitiveArray::patch` require `ArrowNativeType`...? feels weird to leak that into our API

Reply:
This is because our buffers are arrow buffers and to get mutable arrow buffer we need `ArrowNativeType` of said buffer